// react-dnd/lib/esm/TargetConnector.js
import wrapConnectorHooks from './wrapConnectorHooks';
import { isRef } from './utils/isRef';
const shallowEqual = require('shallowequal');
export default class TargetConnector {
    constructor(backend) {
        this.backend = backend;
        this.hooks = wrapConnectorHooks({
            dropTarget: (node, options) => {
                this.dropTargetOptions = options;
                if (isRef(node)) {
                    this.dropTargetRef = node;
                }
                else {
                    this.dropTargetNode = node;
                }
                this.reconnect();
            },
        });
        this.handlerId = null;
        // The drop target may either be attached via ref or connect function
        this.dropTargetRef = null;
        this.dropTargetOptionsInternal = null;
        this.lastConnectedHandlerId = null;
        this.lastConnectedDropTarget = null;
        this.lastConnectedDropTargetOptions = null;
    }
    get connectTarget() {
        return this.dropTarget;
    }
    reconnect() {
        const dropTarget = this.dropTarget;
        if (!this.handlerId || !dropTarget) {
            return;
        }
        // if nothing has changed then don't resubscribe
        if (this.didHandlerIdChange() ||
            this.didDropTargetChange() ||
            this.didOptionsChange()) {
            this.disconnectDropTarget();
            this.lastConnectedHandlerId = this.handlerId;
            this.lastConnectedDropTarget = dropTarget;
            this.lastConnectedDropTargetOptions = this.dropTargetOptions;
            this.unsubscribeDropTarget = this.backend.connectDropTarget(this.handlerId, dropTarget, this.dropTargetOptions);
        }
    }
    receiveHandlerId(newHandlerId) {
        if (newHandlerId === this.handlerId) {
            return;
        }
        this.handlerId = newHandlerId;
        this.reconnect();
    }
    get dropTargetOptions() {
        return this.dropTargetOptionsInternal;
    }
    set dropTargetOptions(options) {
        this.dropTargetOptionsInternal = options;
    }
    didHandlerIdChange() {
        return this.lastConnectedHandlerId !== this.handlerId;
    }
    didDropTargetChange() {
        return this.lastConnectedDropTarget !== this.dropTarget;
    }
    didOptionsChange() {
        return !shallowEqual(this.lastConnectedDropTargetOptions, this.dropTargetOptions);
    }
    disconnectDropTarget() {
        if (this.unsubscribeDropTarget) {
            this.unsubscribeDropTarget();
            this.unsubscribeDropTarget = undefined;
        }
    }
    get dropTarget() {
        return (this.dropTargetNode || (this.dropTargetRef && this.dropTargetRef.current));
    }
}
